### PR TITLE
Draw String With Custom Charset And Custom Fonts.

### PR DESCRIPTION
**Pull request by nifanfa**
QQ截图20210621202723
****
QQ截图20210621195934
# It support Anti-Aliasing(not real) and it's fast. Don't worry about the performance! 
****
QQ截图20210621194656
👉[Generator Here](https://github.com/CosmosOS/Cosmos/files/6686362/Debug.zip)
👉[Source](https://github.com/nifanfa/BitFont)
****
### Actually I want add the generator here but i don't know how to modify it...
QQ截图20210621195205
****
QQ截图20210621195802

### CODE REVIEW

**Comment by nifanfa** [comment]
QQ截图20210621202723

**Comment by nifanfa** [comment]
By the way. i think this problem must be solved #1844

**Comment by AcaiBerii** [comment]
How do you use it?

**Comment by nifanfa** [comment]
> How do you use it?

image

**Comment by elkaamee326** [comment]
Does it support the `DoubleBufferedVMwareSVGAII` driver?

**Comment by nifanfa** [comment]
> Does it support the `DoubleBufferedVMwareSVGAII` driver / canvas?

it only support canvas. but if you want you can implement it!

**Comment by nifanfa** [comment]
it just uses DrawPoint()

File: source/Cosmos.System2/Graphics/Fonts/BitFont.cs [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+        public BitFontDescriptor(string aCharset, MemoryStream aMS, int aSize)
+        {
+            this.Charset = aCharset;
+            this.MS = aMS;
+            this.Size = aSize;
+        }
+    }
+
+    /// <summary>
+    /// For More https://github.com/nifanfa/BitFont
```
Comment (Arawn-Davies) [inline source/Cosmos.System2/Graphics/Fonts/BitFont.cs:22]: Please don't redirect users to another repo.
 If you want to add further functionality into Cosmos, please update/create an article in the Docs/ subdirectory.

Comment (nifanfa) [reply source/Cosmos.System2/Graphics/Fonts/BitFont.cs:22]: sure. i just forgot to remove it.

**Review by terminal-cs — APPROVED** [review]
This new method is very fast and has no downsides, it works very well.

**Comment by quajak** [comment]
Thanks for the PR. It would be very cool to have better font support in Cosmos. Just a few questions:
- Is there any specification for the format of the BitFont? It would be very good if you could add some comment, so that other people know how to create a BitFont.
- When implementing the other font support for Cosmos, I created the abstract class https://github.com/CosmosOS/Cosmos/blob/master/source/Cosmos.System2/Graphics/Fonts/Font.cs would it be possible to make BitFont implement it? Do you think the interface should be changed?
- For the drawing methods, could they also be part of the Canvas class directly? That would keep the interface consistent. Even better would be if drawing strings using BitFont or PCScreenFont can be done using the same method.

**Comment by terminal-cs** [comment]
i think im noticing an issue with it actualy, the font looks "wobily" and uneven, you can see it in this image, you need to properly align the fonts, but from figure 3 it looks ok
Screenshot 2021-07-16 151904
Screenshot 2021-07-16 151920
Screenshot 2021-07-16 151848

**Comment by terminal-cs** [comment]
nianfa, could you please respond to quajak plez? this is cool and should be implemented.

**Comment by nifanfa** [comment]
> nianfa, could you please respond to quajak plez? this is cool and should be implemented.

Yes. But in fact, there is no difference between bitfont in Mosa and Cosmos

**Comment by nifanfa** [comment]
> i think im noticing an issue with it actualy, the font looks "wobily" and uneven, you can see it in this image, you need to properly align the fonts, but from figure 2 it looks ok Screenshot 2021-07-16 151904 Screenshot 2021-07-16 151920 Screenshot 2021-07-16 151848

That

**Comment by nifanfa** [comment]
> i think im noticing an issue with it actualy, the font looks "wobily" and uneven, you can see it in this image, you need to properly align the fonts, but from figure 2 it looks ok Screenshot 2021-07-16 151904 Screenshot 2021-07-16 151920 Screenshot 2021-07-16 151848

You can simply modify the generator

**Comment by terminal-cs** [comment]
Cool, finaly reopened, but I think this needs to be refined a bit

**Comment by RSTPlays** [comment]
How do I use it? It’s a bit confusing and will it support the command line?

**Comment by Neil-Ciciglycerin** [comment]
Why did you delete the repo?

**Comment by Neil-Ciciglycerin** [comment]
Luckily I already had this merged!

**Pull request closed without merging**